### DISCUSSION TITLE
How to configure dev container for graphical apps?

### DISCUSSION THREAD

**Discussion by sliekens** [devcontainers/community#95] [Q&A]
Hello,

I am wondering what is the intended way to configure a devcontainer to run graphical apps? I noticed that VSCode supports automatically mounting a Wayland socket. But there is no easy tutorial that I can find.

I use `mcr.microsoft.com/devcontainers/base:ubuntu` and I believe I got X11 forwarding to work by installing a bunch of X11 packages, but is that really the right way to do it?

``` sh
sudo apt-get install -y pkg-config libasound2-dev libudev-dev mesa-utils vulkan-tools libwayland-dev libxkbcommon-dev libvulkan1 libvulkan-dev libegl1-mesa-dev libgles2-mesa-dev libx11-dev libxcursor-dev libxrandr-dev libxi-dev libxrandr-dev libxcb1-dev libxcb-icccm4-dev libxcb-image0-dev libxcb-keysyms1-dev libxcb-randr0-dev libxcb-shape0-dev libxcb-xfixes0-dev libxcb-xkb-dev libegl1-mesa libgl1-mesa-glx libgl1-mesa-dri libglu1-mesa-dev libglu1-mesa libgles2-mesa
```

I found this list of packages somewhere online and I've no idea what's what... but it worked.

Seems like there's got to be a better way to do it? Using a devcontainer feature?

Thanks.

**Comment by chrmarti** [comment]
I would expect any UI app to have its dependencies set up this way. E.g., when you install `gedit` with a package manager all the required packages you list should automatically get installed too. This should allow for using the X11 and Wayland displays with the Dev Containers extension.

There is also the `ghcr.io/devcontainers/features/desktop-lite` feature that will also install VNC and noVNC, so you could get the UI apps displayed in a local VNC client or a browser.

Are you looking for a way to use graphical apps locally with the Dev Containers extension or also with Codespaces?

**Reply by sliekens** [reply]
In my case I was trying to run a Java Swing app which I built from source. I only had JDK installed in the devcontainer base image, so it didn't work because of missing dependencies (X11 stuff) and it wasn't really obvious what I had to do to get it working. I got it to work thanks to a blog post I found but I don't really know if installing all those packages was the right thing to do.

I don't plan to run GUI apps in codespaces, I would only compile the app in codespaces and then download it to run on my machine.

**Reply by chrmarti** [reply]
I see, so you might indeed need the `-dev` packages you list to develop (not only run) the graphical app. Which ones exactly you need depends on the compile-time and runtime dependencies your development setup has. Not being an expert on this, I would probably try to compile and run it with various subsets of packages to find the minimal list of packages.
